File: h5features2/item.py
from copy import deepcopy
import numpy as np
from pyh5features import Item as pyitem

class Item:
    """This class implement the interface with the python wrapper Item from h5features2

    It allow to create an object with several characteristics
    The features are datas of type time x characteristics. They store several value each of several segment, for example of time
    The time are datas of type time x (begin-end). The store the begin and the end of each time segment
    The properties are datas stored in an hash table. They contribute to characterize the Item
    """
    def __init__(self, item):
        """Constructor of Item class
        
        Args:
            item (`pyh5features.Item`): Construct directly Item from wrapper
                It is used internaly.
        Raises:
            TypeError: If `item` is not an instance of `pyh5features.Item`
        """
        if not isinstance(item, pyitem):
            raise TypeError("item must have pyh5features.Item class")
        self.__item = item
    @classmethod
    def create(cls, name, features, times, properties={}):
        """Main constructor of Item class for user usage

        Args:
            name (`str`): a string to qualify the `Item`.

            features (`numpy.ndarray`): a two-dimensionnal numpy array of data.
                The first dimension correspond to times or segments in the data.
                The second dimension correspond to one/several values according to shape for the segment.

            times (`tuple` of `numpy.ndarray`): a tuple with two numpy array.
                The first correspond to a numpy array where each value correspond to the start time of same index in first dimension of features.
                The second correspond to end time in the same way.

            properties (`dict`, optional): a python dictionnary to record characteristics on the Item.


        Notes:
            properties values can be either `dict`, `bool`, `int`, `float`, `string`, `list` of `int`, `float`, `string`

        Returns:
            Item

        Raises:
            TypeError: if item's name is not `str`, if features is not a two dimensionnal `numpy.ndarray` of `float64 type,
                if times is not a `tuple` of two  one-dimensionnal `numpy.ndarray` of `float64` type,
                if properties keys are not `str`.

        """
        def rec_properties(props):
            for k, v in props.items():
                if not isinstance(k, str):
                    raise TypeError("propertie key must be str")
                if isinstance(v, dict):
                    rec_properties(v)

        if not isinstance(times, tuple):
            raise TypeError("times is not a tuple")
        if len(times) != 2:
            raise TypeError("times must contain to numpy arrays")
        start, stop = times
        if not isinstance(name, str):
            raise TypeError("item's name must be str")
        if not isinstance(features, np.ndarray):
            raise TypeError("features is not a np.ndarray")
        if features.dtype != "float64":
            raise TypeError("features must have float64 type")
        if len(features.shape) != 2:
            raise TypeError("features must be two-dimensionnal")
        if not isinstance(start, np.ndarray):
            raise TypeError("start is not a np.ndarray")
        if start.dtype != "float64":
            raise TypeError("start must have float64 type")
        if not isinstance(stop, np.ndarray):
            raise TypeError("stop is not a np.ndarray")
        if stop.dtype != "float64":
            raise TypeError("stop must have float64 type") 
        if len(start.shape) != 1:
            raise TypeError("start must be one-dimensionnal")
        if len(stop.shape) != 1:
            raise TypeError("stop must be one-dimensionnal") 
        if not isinstance(properties, dict):
            raise TypeError("properties is not a dict")
        rec_properties(properties)

        return cls(pyitem(name, features, start, stop, properties, True))


    def __eq__(self, it) -> bool:
        """Test equality of two Item

        the method test the equality of features, times and properties from the wrapper

        Returns:
            bool: True if equal, else False
        """
        return self.__item == it
    def __ne__(self, it):
        """Test inequality of two Item

        the method test the inequality of features, times and properties from the wrapper

        Returns:
            bool: True if inequal, else False
        """
        return self.__item != it

    def features(self,  copy=False) -> np.ndarray:
        """The method returns the features of the Item

        Allow to return the reference or to do a deepcopy of the features

        Args:
            copy (`bool`): If True, return a deepcopy of the features
                else, return the reference

        Returns:
           np.ndarray : the features

        """
        if copy:
            return deepcopy(np.asarray(self.__item.features(), dtype=np.float64))
        return np.array(self.__item.features(),dtype=np.float64,  copy=False)

    def times(self, copy=False) -> np.ndarray:
        """The method returns the times of the Item

        Allow to return the reference or to do a deepcopy of the times

        Args:
            copy (`bool`): If True, return a deepcopy of the times
                else, return the reference

        Returns:
           np.ndarray : the times

        Notes:
            the times returned is two dimensionnal. The first dimension correspond to the segments according to the features
                the second dimension correspond to two values, the start and the stop of the segment

        """
        if copy:
            return deepcopy(np.asarray(self.__item.times(), dtype=np.float64))
        return np.array(self.__item.times(), dtype=np.float64, copy=False)
    
    def properties(self) -> dict:
        """This method returns the properties of the Item

        Returns:
            dict : the properties
        """
        return self.__item.properties()
    
    
    def set_properties(self, name, value):
        """This method allow to set a new propertie or update an existing propertie

        Args:
            name (`str`): key of propertie
            value : value of properties, see constructor to check the possible types

        Raises:
            TypeError is name is not `str`
        """
        if not isinstance(name, str):
            raise TypeError("name must be str")
        self.__item.properties_set(name, value)

    def has_properties(self) -> bool:
        """This method allow to check if Item has properties

        Returns:
            bool:True if Itme has properties, else False

        """
        return self.__item.has_properties()
    
    def properties_contains(self, name) -> bool:
        """This method allow to check if a propertie exists

        Args:
            name (`str`) : key of propertie

        Raises:
            TypeError is name is not `str`

        Returns:
            bool : True if contains the name, else False
        """
        if not isinstance(name, str):
            raise TypeError("name must be str")
        return self.__item.properties_contains(name)

    def properties_erase(self, name):
        """This method allow to delete a propertie

        Args:
            name (`str`) : key of propertie

        Raises:
            TypeError is name is not `str`
        """
        if not isinstance(name, str):
            raise TypeError("name must be str")
        self.__item.properties_erase(name)

    def ncharacteristic(self) -> int:
        """ This method returns the number of characterics of one segment of features
        
        Returns:
            int: the number of characteristics
        """
        return self.__item.dim()

    def size(self) -> int:
        """This methods the number of segments of features or times

        Returns:
            int: the number of segments

        """
        return self.__item.size()

File: python_wrapper/src/item.cpp
#include  <item.h>

pybind11::dict return_props(h5features::properties src);

pybind::item::item(const std::string& name,
            const h5features::features& features,
            const h5features::times& times,
            const h5features::properties& properties,
            bool check) : h5features::item::item(name, features, times, properties, check)
{
   
}
pybind::item::item(h5features::item item) : pybind::item(item.name(), item.features(), item.times(), item.properties(), true)
{
   
}

std::string pybind::item::name()  {return h5features::item::name(); }
bool pybind::item::has_properties()  {return h5features::item::has_properties();} 
double pybind::item::dim()  {return h5features::item::dim();} 
double pybind::item::size() {return h5features::item::size();} 
bool pybind::item::operator==(const pybind::item& other) const noexcept {return h5features::item::operator==(other);} 
         /// Returns true if the two items are not equal
bool pybind::item::operator!=(const pybind::item& other) const noexcept {return h5features::item::operator!=(other);} 

bool pybind::item::properties_contains(const std::string& name)
{
   /* check if properties exists*/
   return h5features::item::properties().contains(name);
}
void pybind::item::erase_properties(const std::string& name)
{
   /*delete a propertie*/
   h5features::properties& ref = const_cast <h5features::properties&>(h5features::item::properties());
   ref.erase(name);
}


void pybind::item::set_properties(const std::string& name, pybind11::handle src)
{
   /* parse properties from python*/
   PyObject* pyob = src.ptr();
      if (pybind11::isinstance<pybind11::bool_>(src)){
         
         bool value = pybind11::handle(pyob).cast<bool>();
         h5features::properties& ref = const_cast <h5features::properties&>(h5features::item::properties());
         ref.set<decltype(value)>(name, value);
      }
      else if (pybind11::isinstance<pybind11::int_>(src)){
         
         int value = pybind11::handle(pyob).cast<int>();
         h5features::properties& ref = const_cast <h5features::properties&>(h5features::item::properties());
         ref.set<decltype(value)>(name, value);
      }
      else if (pybind11::isinstance<pybind11::float_>(src)){
         double value = pybind11::handle(pyob).cast<double>();
         h5features::properties& ref = const_cast <h5features::properties&>(h5features::item::properties());
         ref.set<decltype(value)>(name, value);
      }
      else if (pybind11::isinstance<pybind11::str>(src)){
         std::string value = pybind11::handle(pyob).cast<std::string>();
         h5features::properties& ref = const_cast <h5features::properties&>(h5features::item::properties());
         ref.set<decltype(value)>(name, value);
      }
      else if (pybind11::isinstance<pybind11::list>(src)){
         if (pybind11::isinstance<pybind11::str>(*src.begin())) {
            std::vector<std::string> value = pybind11::handle(pyob).cast<std::vector<std::string>>();
            h5features::properties& ref = const_cast <h5features::properties&>(h5features::item::properties());
            ref.set<decltype(value)>(name, value);
         }
         else if (pybind11::isinstance<pybind11::int_>(*src.begin())) {
            std::vector<int> value = pybind11::handle(pyob).cast<std::vector<int>>();
            h5features::properties& ref = const_cast <h5features::properties&>(h5features::item::properties());
            ref.set<decltype(value)>(name, value);
         }
         else if (pybind11::isinstance<pybind11::float_>(*src.begin())) {
            std::vector<double> value = pybind11::handle(pyob).cast<std::vector<double>>();
            h5features::properties& ref = const_cast <h5features::properties&>(h5features::item::properties());
            ref.set<decltype(value)>(name, value);
         }
      }
      else if (pybind11::isinstance<pybind11::dict>(src)){ 

         h5features::properties value = pybind11::handle(pyob).cast<h5features::properties>();
         h5features::properties& ref = const_cast <h5features::properties&>(h5features::item::properties());
         ref.set<decltype(value)>(name, value);
   }
   
}


pybind11::dict return_props(h5features::properties src)
{
   /* parse properties type from cpp*/
      pybind11::dict to_return;
      for (auto item : src)
      {
         if (item.second.type().name() == typeid(int).name()){
            to_return[pybind11::str(PyUnicode_DecodeUTF8(item.first.data(), item.first.length(), NULL))] = pybind11::int_(boost::get<int>(item.second));}
         else if (item.second.type().name() == typeid(bool).name())
            to_return[pybind11::str(PyUnicode_DecodeUTF8(item.first.data(), item.first.length(), NULL))] = pybind11::bool_(boost::get<bool>(item.second));
         else if (item.second.type().name() ==typeid(double).name()){
            to_return[pybind11::str(PyUnicode_DecodeUTF8(item.first.data(), item.first.length(), NULL))] = pybind11::float_(boost::get<double>(item.second));
            }
         else if (item.second.type().name() ==typeid(std::string).name()){
            std::string str = boost::get<std::string>(item.second);
            to_return[pybind11::str(PyUnicode_DecodeUTF8(item.first.data(), item.first.length(), NULL))] = pybind11::str(PyUnicode_DecodeUTF8(str.data(), str.length(), NULL));
         }
         else if (item.second.type() ==typeid(std::vector<int>))
            to_return[pybind11::str(PyUnicode_DecodeUTF8(item.first.data(), item.first.length(), NULL))] = pybind11::cast(boost::get<std::vector<int>>(item.second));
         else if (item.second.type() ==typeid(std::vector<double>))
            to_return[pybind11::str(PyUnicode_DecodeUTF8(item.first.data(), item.first.length(), NULL))] = pybind11::cast(boost::get<std::vector<double>>(item.second));
         else if (item.second.type() ==typeid(std::vector<std::string>)) {
            to_return[pybind11::str(PyUnicode_DecodeUTF8(item.first.data(), item.first.length(), NULL))] = pybind11::cast(boost::get<std::vector<std::string>>(item.second));
         }
         else if  (item.second.type() == typeid(h5features::properties)) {
            to_return[pybind11::str(PyUnicode_DecodeUTF8(item.first.data(), item.first.length(), NULL))] = return_props(boost::get<h5features::properties>(item.second));
         }
   }
      return to_return;
}


pybind11::dict pybind::item::properties()
{
   /* return properties */
   auto src = h5features::item::properties();
   return return_props(src);
}

pybind11::array_t<double> pybind::item::features()
{

   /*
   return features
   */
   double* p=(double*)h5features::item::features().data().data();
         return  pybind11::array_t<double>(
            {h5features::item::features().size(),h5features::item::features().dim()}, p
            , 
            pybind11::capsule(
        new auto(p)
        ,
        [](void* ptr){ delete reinterpret_cast<decltype(p)*>(ptr); }
    )
    );
}

pybind11::array_t<double> pybind::item::times()
{
   /*
   return an array of time with row, the number of times, and columns, the start and end of time
   */
   double* p=(double*)h5features::item::times().data().data();
         return  pybind11::array_t<double>(
            {h5features::item::times().size(),h5features::item::times().dim()}, p
            , pybind11::capsule(
        new auto(p),
        [](void* ptr){ delete reinterpret_cast<decltype(p)*>(ptr); }
    )
    );
}